### ISSUE TITLE
use_effect cleanup with use_state

### ISSUE THREAD

**Issue by jamesdbrock** [pyedifice/pyedifice#116]
Scenario: We have a component with a `use_effect` with a `cleanup` function.

When the component unmounts, we want the `cleanup` function to do something with the current `use_state` of the component.

But the current `use_state` is not available to the `cleanup` function, because the `cleanup` function was closed over the `use_state` variables when the `use_effect` was last called.

**Comment by jamesdbrock** [comment]
If we have a `use_effect` that always runs on every render, then the `cleanup` function will be closed on the latest `use_state` variables.

But currently it's hard to make a `use_effect` that always runs. We could change `use_effect` so that if the deps are `None` then the effect always runs. We could default deps to `None`. That is how React's `useEffect` works. https://react.dev/reference/react/useEffect#parameters

**Comment by jamesdbrock** [comment]
`use_effect` always runs with deps `None` https://github.com/pyedifice/pyedifice/commit/0967cdfbffdb1caa7980325b37ca693cc359a068

**Comment by jamesdbrock** [comment]
Hang on this won't give us an unmount cleanup function because with `deps` `None` the cleanup function will run on every render.

**Comment by jamesdbrock** [comment]
I guess React has the same problem https://stackoverflow.com/questions/68124694/useeffect-cleanup-on-unmount-with-dependencies

**Comment by jamesdbrock** [comment]
I wrote and tested this Hook and it works. I'm thinking maybe I should add this to Edifice.

```python
def use_unmount(cleanup: Callable[[], None]):
    """
    When component unmounts, call the :code:`cleanup` function with the latest
    local state from :func:`use_state` Hooks.

    This Hook solves the problem of using :func:`use_effect` with constant
    deps to run a :code:`cleanup`
    function when a component unmounts. If the :func:`use_effect` deps
    are constant so that the :code:`cleanup` function only runs once, then the
    :code:`cleanup` function will not have a closure on the latest state.

    Example::

        x, set_x = ed.use_state(0)

        def unmount_cleanup_x():
            print(f"At unmount, the value of x is {x}")

        use_unmount(unmount_cleanup_x, [x])

    """
    internal_mutable, _ = ed.use_state(cast(list[Callable[[],None]], []))

    # Always re-bind the cleanup function closed on the latest state
    def bind_cleanup():
        if len(internal_mutable) == 0:
            internal_mutable.append(cleanup)
        else:
          internal_mutable[0] = cleanup
    ed.use_effect(bind_cleanup, None)

    # This unmount function is called when the component unmounts
    def unmount():
        def internal_cleanup():
            internal_mutable[0]()
        return internal_cleanup
    ed.use_effect(unmount, [])
```

**Comment by jamesdbrock** [comment]
Maybe `use_unmount` should take a `deps` so it can be used to clean up things other than components.

**Comment by jamesdbrock** [comment]
Published as `use_effect_final` in #134
